app/events/services/service_deleted_event.rb

```diff
@@ -16,4 +16,8 @@ def self.create(service)
 
     new(data)
   end
+
+  def self.valid?(service)
+    EventStore::Event.where(event_type: self).where('data LIKE ?', "%service_id: #{service.id}%").none?
```

Comment:
Hopefully `PurgeOldEventsWorker` will run and this can be less costy.

Reply:
Hmmm I'd avoid doing that for events.
Preferably better spot where it is created multiple times.

Reply:
This has nothing to do with cleaning up 😄 The reason is we have this bug for `BackendDeleteEndUsersWorker`:
```
ThreeScale::Core::APIClient::APIError: ThreeScale::Core::APIClient::APIError: 400 DELETE /internal/services/{:service_id}/users, attributes: {:status=>"error", :error=>"User requires a valid service, t...
```

The reason this happens is explained and shown in https://issues.jboss.org/browse/THREESCALE-2031?focusedCommentId=13705515&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13705515
Please read that comment 😄

Reply:
@hallelujah I prefer that too and I really tried to figure out where that happens and I have no idea! 😞

Reply:
@hallelujah The only place I found is:
https://github.com/3scale/porta/blob/aedd21649ee5d4dce994c92a02e93bfbda3cb6db/app/observers/service_observer.rb#L10-L12
And I have no idea why that would be executed twice 😕

Reply:
@hallelujah 's theory, although he is not sure:

> OK it could be one thing
> it is called inside a transaction
> and the transaction is reverted
> then the service is scheduled for deletion again
> we should call that when and only when the service is deleted for real
> after commit
> not after destroy :)

I will try with that knowledge then 😉

Reply:
@Martouta 
I have a theory but I doubt it, though it might be worth investigating.

We have this https://issues.jboss.org/browse/THREESCALE-1194. Basically it tries to destroy a service in a worker. If it can't be achieved (because of race condition) then it will reschedule the same job.

What happens here, is that `after_destroy` runs inside a transaction. So actually if it fails to destroy as the object, the transaction is rolled back BUT ...

It might be that observers are not run inside this transaction, making it creating the delete event
if that is the case then we should get rid of observers. Anyway they are deprecated.

Reply:
@hallelujah With the code that we have in master, if I write this test in `ServiceObserverTest`, it passes:
```ruby
test 'Services::ServiceDeletedEvent is created after the destroy transaction only' do
    service = FactoryBot.create(:simple_service)
    service.stubs(destroyable?: true)
    assert_no_difference(EventStore::Event.where(event_type: Services::ServiceDeletedEvent).method(:count)) do
      Service.transaction do
        service.destroy!
        raise ActiveRecord::Rollback
      end
    end
    assert_difference(EventStore::Event.where(event_type: Services::ServiceDeletedEvent).method(:count)) do
      service.reload.destroy!
    end
end
```

Reply:
(I'm still investigating, don't worry 😄 )